diff -up openssh-5.8p1/acss.c.audit2a openssh-5.8p1/acss.c
diff -up openssh-5.8p1/acss.h.audit2a openssh-5.8p1/acss.h
diff -up openssh-5.8p1/addrmatch.c.audit2a openssh-5.8p1/addrmatch.c
diff -up openssh-5.8p1/atomicio.c.audit2a openssh-5.8p1/atomicio.c
diff -up openssh-5.8p1/atomicio.h.audit2a openssh-5.8p1/atomicio.h
diff -up openssh-5.8p1/audit-bsm.c.audit2a openssh-5.8p1/audit-bsm.c
--- openssh-5.8p1/audit-bsm.c.audit2a	2011-02-21 16:17:09.000000000 +0100
+++ openssh-5.8p1/audit-bsm.c	2011-02-21 16:20:57.000000000 +0100
@@ -317,7 +317,7 @@ audit_session_close(struct logininfo *li
 }
 
 int
-audit_keyusage(int host_user, const char *type, unsigned len, char *fp, int rv)
+audit_keyusage(int host_user, const char *type, unsigned bits, char *fp, int rv)
 {
 	/* not implemented */
 }
diff -up openssh-5.8p1/audit.c.audit2a openssh-5.8p1/audit.c
--- openssh-5.8p1/audit.c.audit2a	2011-02-21 16:17:09.000000000 +0100
+++ openssh-5.8p1/audit.c	2011-02-21 16:23:39.000000000 +0100
@@ -36,6 +36,7 @@
 #include "key.h"
 #include "hostfile.h"
 #include "auth.h"
+#include "xmalloc.h"
 
 /*
  * Care must be taken when using this since it WILL NOT be initialized when
@@ -111,29 +112,18 @@ audit_event_lookup(ssh_audit_event_t ev)
 	return(event_lookup[i].name);
 }
 
-int
-audit_key(int type, int *rv, const Key *key)
+void
+audit_key(int host_user, int *rv, const Key *key)
 {
 	char *fp;
-	unsigned size = 0;
-	const char *crypto_name[] = {
-		"ssh-rsa1",
-		"ssh-rsa",
-		"ssh-dsa",
-		"unknown" };
+	const char *crypto_name;
 
 	fp = key_fingerprint(key, SSH_FP_MD5, SSH_FP_HEX);
-	switch(key->type) {
-		case KEY_RSA1:
-		case KEY_RSA:
-			size = RSA_size(key->rsa);
-			break;
-		case KEY_DSA:
-			size = DSA_size(key->dsa);
-			break;
-	}
-
-	if (audit_keyusage(0, crypto_name[key->type], size, fp, *rv) == 0)
+	if (key->type == KEY_RSA1)
+		crypto_name = "ssh-rsa1";
+	else
+		crypto_name = key_ssh_name(key);
+	if (audit_keyusage(host_user, crypto_name, key_size(key), fp, *rv) == 0)
 		*rv = 0;
 	xfree(fp);
 }
@@ -216,10 +206,10 @@ audit_run_command(const char *command)
  * Type is the key type, len is the key length(byte) and fp is the fingerprint of the key.
  */
 int
-audit_keyusage(int host_user, const char *type, unsigned len, char *fp, int rv)
+audit_keyusage(int host_user, const char *type, unsigned bits, char *fp, int rv)
 {
 	debug("audit %s key usage euid %d user %s key type %s key length %d fingerprint %s, result %d", 
-		host_user ? "hostbased" : "pubkey", geteuid(), audit_username(), type, len, fp, rv);
+		host_user ? "hostbased" : "pubkey", geteuid(), audit_username(), type, bits, fp, rv);
 }
 # endif  /* !defined CUSTOM_SSH_AUDIT_EVENTS */
 #endif /* SSH_AUDIT_EVENTS */
diff -up openssh-5.8p1/audit.h.audit2a openssh-5.8p1/audit.h
--- openssh-5.8p1/audit.h.audit2a	2011-02-21 16:17:09.000000000 +0100
+++ openssh-5.8p1/audit.h	2011-02-21 16:24:27.000000000 +0100
@@ -55,6 +55,6 @@ void	audit_session_close(struct logininf
 void	audit_run_command(const char *);
 ssh_audit_event_t audit_classify_auth(const char *);
 int	audit_keyusage(int, const char *, unsigned, char *, int);
-int	audit_key(int, int *, const Key *);
+void	audit_key(int, int *, const Key *);
 
 #endif /* _SSH_AUDIT_H */
diff -up openssh-5.8p1/audit-linux.c.audit2a openssh-5.8p1/audit-linux.c
--- openssh-5.8p1/audit-linux.c.audit2a	2011-02-21 16:17:09.000000000 +0100
+++ openssh-5.8p1/audit-linux.c	2011-02-21 16:21:19.000000000 +0100
@@ -129,7 +129,7 @@ fatal_report:
 }
 
 int
-audit_keyusage(int host_user, const char *type, unsigned len, char *fp, int rv)
+audit_keyusage(int host_user, const char *type, unsigned bits, char *fp, int rv)
 {
 	char buf[AUDIT_LOG_SIZE];
 	int audit_fd, rc, saved_errno;
@@ -148,7 +148,7 @@ audit_keyusage(int host_user, const char
 	if ((rc < 0) && ((rc != -1) || (getuid() == 0)))
 		goto out;
 	snprintf(buf, sizeof(buf), "key algo=%s size=%d fp=%s rport=%d",
-			type, 8 * len, fp, get_remote_port());
+			type, bits, fp, get_remote_port());
 	rc = audit_log_acct_message(audit_fd, AUDIT_USER_AUTH, NULL,
 		buf, audit_username(), -1, NULL, get_remote_ipaddr(), NULL, rv);
 out:
diff -up openssh-5.8p1/auth1.c.audit2a openssh-5.8p1/auth1.c
diff -up openssh-5.8p1/auth2.c.audit2a openssh-5.8p1/auth2.c
diff -up openssh-5.8p1/auth2-chall.c.audit2a openssh-5.8p1/auth2-chall.c
diff -up openssh-5.8p1/auth2-gss.c.audit2a openssh-5.8p1/auth2-gss.c
diff -up openssh-5.8p1/auth2-hostbased.c.audit2a openssh-5.8p1/auth2-hostbased.c
diff -up openssh-5.8p1/auth2-jpake.c.audit2a openssh-5.8p1/auth2-jpake.c
diff -up openssh-5.8p1/auth2-kbdint.c.audit2a openssh-5.8p1/auth2-kbdint.c
diff -up openssh-5.8p1/auth2-none.c.audit2a openssh-5.8p1/auth2-none.c
diff -up openssh-5.8p1/auth2-passwd.c.audit2a openssh-5.8p1/auth2-passwd.c
diff -up openssh-5.8p1/auth2-pubkey.c.audit2a openssh-5.8p1/auth2-pubkey.c
diff -up openssh-5.8p1/auth-bsdauth.c.audit2a openssh-5.8p1/auth-bsdauth.c
diff -up openssh-5.8p1/auth.c.audit2a openssh-5.8p1/auth.c
diff -up openssh-5.8p1/auth-chall.c.audit2a openssh-5.8p1/auth-chall.c
diff -up openssh-5.8p1/authfd.c.audit2a openssh-5.8p1/authfd.c
diff -up openssh-5.8p1/authfd.h.audit2a openssh-5.8p1/authfd.h
diff -up openssh-5.8p1/authfile.c.audit2a openssh-5.8p1/authfile.c
diff -up openssh-5.8p1/authfile.h.audit2a openssh-5.8p1/authfile.h
diff -up openssh-5.8p1/auth.h.audit2a openssh-5.8p1/auth.h
diff -up openssh-5.8p1/auth-krb5.c.audit2a openssh-5.8p1/auth-krb5.c
diff -up openssh-5.8p1/auth-options.c.audit2a openssh-5.8p1/auth-options.c
diff -up openssh-5.8p1/auth-options.h.audit2a openssh-5.8p1/auth-options.h
diff -up openssh-5.8p1/auth-pam.c.audit2a openssh-5.8p1/auth-pam.c
diff -up openssh-5.8p1/auth-pam.h.audit2a openssh-5.8p1/auth-pam.h
diff -up openssh-5.8p1/auth-passwd.c.audit2a openssh-5.8p1/auth-passwd.c
diff -up openssh-5.8p1/auth-rhosts.c.audit2a openssh-5.8p1/auth-rhosts.c
diff -up openssh-5.8p1/auth-rh-rsa.c.audit2a openssh-5.8p1/auth-rh-rsa.c
diff -up openssh-5.8p1/auth-rsa.c.audit2a openssh-5.8p1/auth-rsa.c
--- openssh-5.8p1/auth-rsa.c.audit2a	2011-02-21 16:17:09.000000000 +0100
+++ openssh-5.8p1/auth-rsa.c	2011-02-21 16:25:17.000000000 +0100
@@ -120,7 +120,7 @@ auth_rsa_verify_response(Key *key, BIGNU
 
 #ifdef SSH_AUDIT_EVENTS
 	fp = key_fingerprint(key, SSH_FP_MD5, SSH_FP_HEX);
-	if (audit_keyusage(1, "ssh-rsa1", RSA_size(key->rsa), fp, rv) == 0) {
+	if (audit_keyusage(1, "ssh-rsa1", RSA_size(key->rsa) * 8, fp, rv) == 0) {
 		debug("unsuccessful audit");
 		rv = 0;
 	}
diff -up openssh-5.8p1/auth-shadow.c.audit2a openssh-5.8p1/auth-shadow.c
diff -up openssh-5.8p1/auth-sia.c.audit2a openssh-5.8p1/auth-sia.c
diff -up openssh-5.8p1/auth-sia.h.audit2a openssh-5.8p1/auth-sia.h
diff -up openssh-5.8p1/auth-skey.c.audit2a openssh-5.8p1/auth-skey.c
diff -up openssh-5.8p1/bufaux.c.audit2a openssh-5.8p1/bufaux.c
diff -up openssh-5.8p1/bufbn.c.audit2a openssh-5.8p1/bufbn.c
diff -up openssh-5.8p1/bufec.c.audit2a openssh-5.8p1/bufec.c
diff -up openssh-5.8p1/buffer.c.audit2a openssh-5.8p1/buffer.c
diff -up openssh-5.8p1/buffer.h.audit2a openssh-5.8p1/buffer.h
diff -up openssh-5.8p1/canohost.c.audit2a openssh-5.8p1/canohost.c
diff -up openssh-5.8p1/canohost.h.audit2a openssh-5.8p1/canohost.h
diff -up openssh-5.8p1/channels.c.audit2a openssh-5.8p1/channels.c
diff -up openssh-5.8p1/channels.h.audit2a openssh-5.8p1/channels.h
diff -up openssh-5.8p1/cipher-3des1.c.audit2a openssh-5.8p1/cipher-3des1.c
diff -up openssh-5.8p1/cipher-acss.c.audit2a openssh-5.8p1/cipher-acss.c
diff -up openssh-5.8p1/cipher-aes.c.audit2a openssh-5.8p1/cipher-aes.c
diff -up openssh-5.8p1/cipher-bf1.c.audit2a openssh-5.8p1/cipher-bf1.c
diff -up openssh-5.8p1/cipher.c.audit2a openssh-5.8p1/cipher.c
diff -up openssh-5.8p1/cipher-ctr.c.audit2a openssh-5.8p1/cipher-ctr.c
diff -up openssh-5.8p1/cipher.h.audit2a openssh-5.8p1/cipher.h
diff -up openssh-5.8p1/cleanup.c.audit2a openssh-5.8p1/cleanup.c
diff -up openssh-5.8p1/clientloop.c.audit2a openssh-5.8p1/clientloop.c
diff -up openssh-5.8p1/clientloop.h.audit2a openssh-5.8p1/clientloop.h
diff -up openssh-5.8p1/compat.c.audit2a openssh-5.8p1/compat.c
diff -up openssh-5.8p1/compat.h.audit2a openssh-5.8p1/compat.h
diff -up openssh-5.8p1/compress.c.audit2a openssh-5.8p1/compress.c
diff -up openssh-5.8p1/compress.h.audit2a openssh-5.8p1/compress.h
diff -up openssh-5.8p1/crc32.c.audit2a openssh-5.8p1/crc32.c
diff -up openssh-5.8p1/crc32.h.audit2a openssh-5.8p1/crc32.h
diff -up openssh-5.8p1/deattack.c.audit2a openssh-5.8p1/deattack.c
diff -up openssh-5.8p1/deattack.h.audit2a openssh-5.8p1/deattack.h
diff -up openssh-5.8p1/defines.h.audit2a openssh-5.8p1/defines.h
diff -up openssh-5.8p1/dh.c.audit2a openssh-5.8p1/dh.c
diff -up openssh-5.8p1/dh.h.audit2a openssh-5.8p1/dh.h
diff -up openssh-5.8p1/dispatch.c.audit2a openssh-5.8p1/dispatch.c
diff -up openssh-5.8p1/dispatch.h.audit2a openssh-5.8p1/dispatch.h
diff -up openssh-5.8p1/dns.c.audit2a openssh-5.8p1/dns.c
diff -up openssh-5.8p1/dns.h.audit2a openssh-5.8p1/dns.h
diff -up openssh-5.8p1/entropy.c.audit2a openssh-5.8p1/entropy.c
diff -up openssh-5.8p1/entropy.h.audit2a openssh-5.8p1/entropy.h
diff -up openssh-5.8p1/fatal.c.audit2a openssh-5.8p1/fatal.c
diff -up openssh-5.8p1/groupaccess.c.audit2a openssh-5.8p1/groupaccess.c
diff -up openssh-5.8p1/groupaccess.h.audit2a openssh-5.8p1/groupaccess.h
diff -up openssh-5.8p1/gss-genr.c.audit2a openssh-5.8p1/gss-genr.c
diff -up openssh-5.8p1/gss-serv.c.audit2a openssh-5.8p1/gss-serv.c
diff -up openssh-5.8p1/gss-serv-krb5.c.audit2a openssh-5.8p1/gss-serv-krb5.c
diff -up openssh-5.8p1/hostfile.c.audit2a openssh-5.8p1/hostfile.c
diff -up openssh-5.8p1/hostfile.h.audit2a openssh-5.8p1/hostfile.h
diff -up openssh-5.8p1/includes.h.audit2a openssh-5.8p1/includes.h
diff -up openssh-5.8p1/jpake.c.audit2a openssh-5.8p1/jpake.c
diff -up openssh-5.8p1/jpake.h.audit2a openssh-5.8p1/jpake.h
diff -up openssh-5.8p1/kex.c.audit2a openssh-5.8p1/kex.c
diff -up openssh-5.8p1/kexdh.c.audit2a openssh-5.8p1/kexdh.c
diff -up openssh-5.8p1/kexdhc.c.audit2a openssh-5.8p1/kexdhc.c
diff -up openssh-5.8p1/kexdhs.c.audit2a openssh-5.8p1/kexdhs.c
diff -up openssh-5.8p1/kexecdh.c.audit2a openssh-5.8p1/kexecdh.c
diff -up openssh-5.8p1/kexecdhc.c.audit2a openssh-5.8p1/kexecdhc.c
diff -up openssh-5.8p1/kexecdhs.c.audit2a openssh-5.8p1/kexecdhs.c
diff -up openssh-5.8p1/kexgex.c.audit2a openssh-5.8p1/kexgex.c
diff -up openssh-5.8p1/kexgexc.c.audit2a openssh-5.8p1/kexgexc.c
diff -up openssh-5.8p1/kexgexs.c.audit2a openssh-5.8p1/kexgexs.c
diff -up openssh-5.8p1/kex.h.audit2a openssh-5.8p1/kex.h
diff -up openssh-5.8p1/key.c.audit2a openssh-5.8p1/key.c
diff -up openssh-5.8p1/key.h.audit2a openssh-5.8p1/key.h
diff -up openssh-5.8p1/log.c.audit2a openssh-5.8p1/log.c
diff -up openssh-5.8p1/log.h.audit2a openssh-5.8p1/log.h
diff -up openssh-5.8p1/loginrec.c.audit2a openssh-5.8p1/loginrec.c
diff -up openssh-5.8p1/loginrec.h.audit2a openssh-5.8p1/loginrec.h
diff -up openssh-5.8p1/logintest.c.audit2a openssh-5.8p1/logintest.c
diff -up openssh-5.8p1/mac.c.audit2a openssh-5.8p1/mac.c
diff -up openssh-5.8p1/mac.h.audit2a openssh-5.8p1/mac.h
diff -up openssh-5.8p1/match.c.audit2a openssh-5.8p1/match.c
diff -up openssh-5.8p1/match.h.audit2a openssh-5.8p1/match.h
diff -up openssh-5.8p1/md5crypt.c.audit2a openssh-5.8p1/md5crypt.c
diff -up openssh-5.8p1/md5crypt.h.audit2a openssh-5.8p1/md5crypt.h
diff -up openssh-5.8p1/md-sha256.c.audit2a openssh-5.8p1/md-sha256.c
diff -up openssh-5.8p1/misc.c.audit2a openssh-5.8p1/misc.c
diff -up openssh-5.8p1/misc.h.audit2a openssh-5.8p1/misc.h
diff -up openssh-5.8p1/moduli.c.audit2a openssh-5.8p1/moduli.c
diff -up openssh-5.8p1/monitor.c.audit2a openssh-5.8p1/monitor.c
diff -up openssh-5.8p1/monitor_fdpass.c.audit2a openssh-5.8p1/monitor_fdpass.c
diff -up openssh-5.8p1/monitor_fdpass.h.audit2a openssh-5.8p1/monitor_fdpass.h
diff -up openssh-5.8p1/monitor.h.audit2a openssh-5.8p1/monitor.h
diff -up openssh-5.8p1/monitor_mm.c.audit2a openssh-5.8p1/monitor_mm.c
diff -up openssh-5.8p1/monitor_mm.h.audit2a openssh-5.8p1/monitor_mm.h
diff -up openssh-5.8p1/monitor_wrap.c.audit2a openssh-5.8p1/monitor_wrap.c
diff -up openssh-5.8p1/monitor_wrap.h.audit2a openssh-5.8p1/monitor_wrap.h
diff -up openssh-5.8p1/msg.c.audit2a openssh-5.8p1/msg.c
diff -up openssh-5.8p1/msg.h.audit2a openssh-5.8p1/msg.h
diff -up openssh-5.8p1/mux.c.audit2a openssh-5.8p1/mux.c
diff -up openssh-5.8p1/myproposal.h.audit2a openssh-5.8p1/myproposal.h
diff -up openssh-5.8p1/nchan.c.audit2a openssh-5.8p1/nchan.c
diff -up openssh-5.8p1/packet.c.audit2a openssh-5.8p1/packet.c
diff -up openssh-5.8p1/packet.h.audit2a openssh-5.8p1/packet.h
diff -up openssh-5.8p1/pathnames.h.audit2a openssh-5.8p1/pathnames.h
diff -up openssh-5.8p1/pkcs11.h.audit2a openssh-5.8p1/pkcs11.h
diff -up openssh-5.8p1/platform.c.audit2a openssh-5.8p1/platform.c
diff -up openssh-5.8p1/platform.h.audit2a openssh-5.8p1/platform.h
diff -up openssh-5.8p1/progressmeter.c.audit2a openssh-5.8p1/progressmeter.c
diff -up openssh-5.8p1/progressmeter.h.audit2a openssh-5.8p1/progressmeter.h
diff -up openssh-5.8p1/readconf.c.audit2a openssh-5.8p1/readconf.c
diff -up openssh-5.8p1/readconf.h.audit2a openssh-5.8p1/readconf.h
diff -up openssh-5.8p1/readpass.c.audit2a openssh-5.8p1/readpass.c
diff -up openssh-5.8p1/rijndael.c.audit2a openssh-5.8p1/rijndael.c
diff -up openssh-5.8p1/rijndael.h.audit2a openssh-5.8p1/rijndael.h
diff -up openssh-5.8p1/roaming_client.c.audit2a openssh-5.8p1/roaming_client.c
diff -up openssh-5.8p1/roaming_common.c.audit2a openssh-5.8p1/roaming_common.c
diff -up openssh-5.8p1/roaming_dummy.c.audit2a openssh-5.8p1/roaming_dummy.c
diff -up openssh-5.8p1/roaming.h.audit2a openssh-5.8p1/roaming.h
diff -up openssh-5.8p1/roaming_serv.c.audit2a openssh-5.8p1/roaming_serv.c
diff -up openssh-5.8p1/rsa.c.audit2a openssh-5.8p1/rsa.c
diff -up openssh-5.8p1/rsa.h.audit2a openssh-5.8p1/rsa.h
diff -up openssh-5.8p1/schnorr.c.audit2a openssh-5.8p1/schnorr.c
diff -up openssh-5.8p1/schnorr.h.audit2a openssh-5.8p1/schnorr.h
diff -up openssh-5.8p1/scp.c.audit2a openssh-5.8p1/scp.c
diff -up openssh-5.8p1/servconf.c.audit2a openssh-5.8p1/servconf.c
diff -up openssh-5.8p1/servconf.h.audit2a openssh-5.8p1/servconf.h
diff -up openssh-5.8p1/serverloop.c.audit2a openssh-5.8p1/serverloop.c
diff -up openssh-5.8p1/serverloop.h.audit2a openssh-5.8p1/serverloop.h
diff -up openssh-5.8p1/session.c.audit2a openssh-5.8p1/session.c
diff -up openssh-5.8p1/session.h.audit2a openssh-5.8p1/session.h
diff -up openssh-5.8p1/sftp.c.audit2a openssh-5.8p1/sftp.c
diff -up openssh-5.8p1/sftp-client.c.audit2a openssh-5.8p1/sftp-client.c
diff -up openssh-5.8p1/sftp-client.h.audit2a openssh-5.8p1/sftp-client.h
diff -up openssh-5.8p1/sftp-common.c.audit2a openssh-5.8p1/sftp-common.c
diff -up openssh-5.8p1/sftp-common.h.audit2a openssh-5.8p1/sftp-common.h
diff -up openssh-5.8p1/sftp-glob.c.audit2a openssh-5.8p1/sftp-glob.c
diff -up openssh-5.8p1/sftp.h.audit2a openssh-5.8p1/sftp.h
diff -up openssh-5.8p1/sftp-server.c.audit2a openssh-5.8p1/sftp-server.c
diff -up openssh-5.8p1/sftp-server-main.c.audit2a openssh-5.8p1/sftp-server-main.c
diff -up openssh-5.8p1/ssh1.h.audit2a openssh-5.8p1/ssh1.h
diff -up openssh-5.8p1/ssh2.h.audit2a openssh-5.8p1/ssh2.h
diff -up openssh-5.8p1/ssh-add.c.audit2a openssh-5.8p1/ssh-add.c
diff -up openssh-5.8p1/ssh-agent.c.audit2a openssh-5.8p1/ssh-agent.c
diff -up openssh-5.8p1/ssh.c.audit2a openssh-5.8p1/ssh.c
diff -up openssh-5.8p1/sshconnect1.c.audit2a openssh-5.8p1/sshconnect1.c
diff -up openssh-5.8p1/sshconnect2.c.audit2a openssh-5.8p1/sshconnect2.c
diff -up openssh-5.8p1/sshconnect.c.audit2a openssh-5.8p1/sshconnect.c
diff -up openssh-5.8p1/sshconnect.h.audit2a openssh-5.8p1/sshconnect.h
diff -up openssh-5.8p1/sshd.c.audit2a openssh-5.8p1/sshd.c
diff -up openssh-5.8p1/ssh-dss.c.audit2a openssh-5.8p1/ssh-dss.c
diff -up openssh-5.8p1/ssh-ecdsa.c.audit2a openssh-5.8p1/ssh-ecdsa.c
diff -up openssh-5.8p1/ssh-gss.h.audit2a openssh-5.8p1/ssh-gss.h
diff -up openssh-5.8p1/ssh.h.audit2a openssh-5.8p1/ssh.h
diff -up openssh-5.8p1/ssh-keygen.c.audit2a openssh-5.8p1/ssh-keygen.c
diff -up openssh-5.8p1/ssh-keyscan.c.audit2a openssh-5.8p1/ssh-keyscan.c
diff -up openssh-5.8p1/ssh-keysign.c.audit2a openssh-5.8p1/ssh-keysign.c
diff -up openssh-5.8p1/sshlogin.c.audit2a openssh-5.8p1/sshlogin.c
diff -up openssh-5.8p1/sshlogin.h.audit2a openssh-5.8p1/sshlogin.h
diff -up openssh-5.8p1/ssh-pkcs11.c.audit2a openssh-5.8p1/ssh-pkcs11.c
diff -up openssh-5.8p1/ssh-pkcs11-client.c.audit2a openssh-5.8p1/ssh-pkcs11-client.c
diff -up openssh-5.8p1/ssh-pkcs11.h.audit2a openssh-5.8p1/ssh-pkcs11.h
diff -up openssh-5.8p1/ssh-pkcs11-helper.c.audit2a openssh-5.8p1/ssh-pkcs11-helper.c
diff -up openssh-5.8p1/sshpty.c.audit2a openssh-5.8p1/sshpty.c
diff -up openssh-5.8p1/sshpty.h.audit2a openssh-5.8p1/sshpty.h
diff -up openssh-5.8p1/ssh-rand-helper.c.audit2a openssh-5.8p1/ssh-rand-helper.c
diff -up openssh-5.8p1/ssh-rsa.c.audit2a openssh-5.8p1/ssh-rsa.c
diff -up openssh-5.8p1/sshtty.c.audit2a openssh-5.8p1/sshtty.c
diff -up openssh-5.8p1/ttymodes.c.audit2a openssh-5.8p1/ttymodes.c
diff -up openssh-5.8p1/ttymodes.h.audit2a openssh-5.8p1/ttymodes.h
diff -up openssh-5.8p1/uidswap.c.audit2a openssh-5.8p1/uidswap.c
diff -up openssh-5.8p1/uidswap.h.audit2a openssh-5.8p1/uidswap.h
diff -up openssh-5.8p1/umac.c.audit2a openssh-5.8p1/umac.c
diff -up openssh-5.8p1/umac.h.audit2a openssh-5.8p1/umac.h
diff -up openssh-5.8p1/uuencode.c.audit2a openssh-5.8p1/uuencode.c
diff -up openssh-5.8p1/uuencode.h.audit2a openssh-5.8p1/uuencode.h
diff -up openssh-5.8p1/version.h.audit2a openssh-5.8p1/version.h
diff -up openssh-5.8p1/xmalloc.c.audit2a openssh-5.8p1/xmalloc.c
diff -up openssh-5.8p1/xmalloc.h.audit2a openssh-5.8p1/xmalloc.h
