diff -up openssh-5.8p2/openbsd-compat/port-linux.c.sesftplog openssh-5.8p2/openbsd-compat/port-linux.c
--- openssh-5.8p2/openbsd-compat/port-linux.c.sesftplog	2011-08-08 12:47:21.000000000 +0200
+++ openssh-5.8p2/openbsd-compat/port-linux.c	2011-08-08 12:54:10.234607372 +0200
@@ -506,7 +506,7 @@ ssh_selinux_change_context(const char *n
 	debug3("%s: setting context from '%s' to '%s'", __func__, oldctx,
 	    newctx);
 	if (setcon(newctx) < 0)
-		logit("%s: setcon failed with %s", __func__, strerror (errno));
+		logit("%s: setcon %s from %s failed with %s", __func__, newctx, oldctx, strerror (errno));
 	xfree(oldctx);
 	xfree(newctx);
 }
