Written-by: Jan Kratochvil <jkratoch@redhat.com>
Reviewed-by: Tomas Mraz <tmraz@redhat.com>

diff -up openssh-4.7p1/channels.h.revert-wsize openssh-4.7p1/channels.h
--- openssh-4.7p1/channels.h.revert-wsize	2007-06-12 15:38:54.000000000 +0200
+++ openssh-4.7p1/channels.h	2007-09-17 23:21:32.000000000 +0200
@@ -122,9 +122,9 @@ struct Channel {
 
 /* default window/packet sizes for tcp/x11-fwd-channel */
 #define CHAN_SES_PACKET_DEFAULT	(32*1024)
-#define CHAN_SES_WINDOW_DEFAULT	(64*CHAN_SES_PACKET_DEFAULT)
+#define CHAN_SES_WINDOW_DEFAULT	(4*CHAN_SES_PACKET_DEFAULT)
 #define CHAN_TCP_PACKET_DEFAULT	(32*1024)
-#define CHAN_TCP_WINDOW_DEFAULT	(64*CHAN_TCP_PACKET_DEFAULT)
+#define CHAN_TCP_WINDOW_DEFAULT	(4*CHAN_TCP_PACKET_DEFAULT)
 #define CHAN_X11_PACKET_DEFAULT	(16*1024)
 #define CHAN_X11_WINDOW_DEFAULT	(4*CHAN_X11_PACKET_DEFAULT)
 
