diff -up openssh-5.8p1/sshd.c.audit5a openssh-5.8p1/sshd.c
--- openssh-5.8p1/sshd.c.audit5a	2011-03-08 09:03:49.000000000 +0100
+++ openssh-5.8p1/sshd.c	2011-03-08 09:06:23.000000000 +0100
@@ -2085,6 +2085,7 @@ main(int ac, char **av)
 
 	/* The connection has been terminated. */
 	packet_destroy_all(1, 1);
+	destroy_sensitive_data(1);
 
 	packet_get_state(MODE_IN, NULL, NULL, NULL, &ibytes);
 	packet_get_state(MODE_OUT, NULL, NULL, NULL, &obytes);
